### PR TITLE
[Docs] Adds external redirects section for App Router

### PR DESCRIPTION
**Pull request by manovotny**
Pages Router [recommended](https://nextjs.org/docs/pages/api-reference/functions/use-router#routerpush:~:text=You%20don%27t%20need,for%20those%20cases) the use of `window.location`.

However, support for external redirects has improved significantly in App Router were we can use `redirect`, `router.push`, etc. natively now.

Slack conversations, for reference: 

* [Initial questions](https://vercel.slack.com/archives/C03S8ED1DKM/p1706021114794229)
* [Continued conversation and experimentation](https://vercel.slack.com/archives/C03S9JCH2Q5/p1706050234819819)

### CODE REVIEW

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 1]
```diff
@@ -252,6 +252,86 @@ export const config = {
 
 See the [Middleware](/docs/app/building-your-application/routing/middleware) documentation for more information.
 
+## External redirects
+
+App Router fully supports programmatic external redirects to other sites (ie. `https://google.com`) using the same methods mentioned above.
```
Comment (leerob) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:257]:
```suggestion
The App Router fully supports programmatic redirects to external links using the same methods mentioned above.
```

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 2]
```diff
@@ -252,6 +252,86 @@ export const config = {
...
+
+An example of when this would be useful is when generating checkout urls for hosted payment providers, such as Stripe and Shopify.
```
Comment (leerob) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:259]: Cut

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 3]
```diff
@@ -252,6 +252,86 @@ export const config = {
...
+
+[`redirect`](#redirect-function) and [permanentRedirect`](#permanentredirect-function) work on the server and in [Server Actions](/docs/app/building-your-application/data-fetching/server-actions-and-mutations).
```
Comment (leerob) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:261]:
```suggestion
[`redirect`](#redirect-function) and [`permanentRedirect`](#permanentredirect-function) work on the server and in [Server Actions](/docs/app/building-your-application/data-fetching/server-actions-and-mutations).
```

What is "on the server" here?

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 4]
```diff
@@ -252,6 +252,86 @@ export const config = {
...
+```ts filename="app/actions.ts" switcher
+'use server'
+
+import { redirect } from 'next/navigation'
+import { createCheckout } from './payment'
+
+export async function checkout(data: FormData) {
+  // Do validation and / or generate url
+  const { url } = await createCheckout(data)
+
```
Comment (leerob) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:272]: I wonder if we even need this, or we just keep the example super brief with the `redirect('https://google.com')` or similar.

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 5]
```diff
@@ -252,6 +252,86 @@ export const config = {
...
+import { useRouter } from 'next/navigation'
+
+export default function Page() {
+  const router = useRouter()
+
+  return (
+    <button
+      type="button"
+      onClick={() => router.push('https://checkout.stripe.com/c/pay')}
```
Comment (leerob) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:304]: Yeah, like here!

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx (hunk 2 above)
Comment (manovotny) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:259]:
```suggestion
```

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 6]
```diff
@@ -252,6 +252,86 @@ export const config = {
 
 See the [Middleware](/docs/app/building-your-application/routing/middleware) documentation for more information.
 
+## External redirects
+
+App Router fully supports programmatic external redirects to other sites (ie. `https://google.com`) using the same methods mentioned above.
+
```
Comment (manovotny) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:258]:
```suggestion
```

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx (hunk 3 above)
Comment (manovotny) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:261]: How about this instead?

```suggestion
[`redirect`](#redirect-function) and [`permanentRedirect`](#permanentredirect-function) work server-side and in [Server Actions](/docs/app/building-your-application/data-fetching/server-actions-and-mutations).
```

Comment (manovotny) [reply docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:259]: How about this instead?

```suggestion
[`redirect`](#redirect-function) and [`permanentRedirect`](#permanentredirect-function) work server-side and in [Server Actions](/docs/app/building-your-application/data-fetching/server-actions-and-mutations).
```

Comment (leerob) [reply docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:259]: SGTM!

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 7]
```diff
@@ -252,6 +252,78 @@ export const config = {
 
 See the [Middleware](/docs/app/building-your-application/routing/middleware) documentation for more information.
 
+## External redirects
```
Comment (delbaoliveira) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:255]:
```suggestion
<AppOnly>

## External redirects
```

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 8]
```diff
@@ -252,6 +252,78 @@ export const config = {
...
+      Checkout
+    </button>
+  )
+}
+```
+
+> **Good to know**:
+>
+> - If you don't need to programmatically navigate a user, you should use a [`<Link>`](/docs/app/api-reference/components/link) component
+
```
Comment (delbaoliveira) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:326]:
```suggestion
</AppOnly>
```

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 9]
```diff
@@ -252,6 +252,78 @@ export const config = {
 
 See the [Middleware](/docs/app/building-your-application/routing/middleware) documentation for more information.
 
+## External redirects
+
+The App Router fully supports programmatic redirects to external links using the same methods mentioned above.
+
+[`redirect`](#redirect-function) and [permanentRedirect`](#permanentredirect-function) work on the server and in [Server Actions](/docs/app/building-your-application/data-fetching/server-actions-and-mutations).
```
Comment (delbaoliveira) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:259]: Can remove the second paragrah imo, since it's mentioned in the overview table. 

```suggestion
`redirect` and `permanentRedirect` can be used for external links using absolute URLs. 
```

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 10]
```diff
@@ -252,6 +252,78 @@ export const config = {
...
+
+import { redirect } from 'next/navigation'
+
+export async function checkout(data) {
+  // Do validation, etc. and then redirect to hosted checkout
+  redirect('https://checkout.stripe.com/c/pay')
+}
+```
+
+`router.push` and `router.replace` from [`useRouter`](#userouter-hook) work in client-side components and event handlers.
```
Comment (delbaoliveira) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:283]:
```suggestion
In event handlers, `router.push` and `router.replace` methods [`useRouter`](#userouter-hook) can be used to navigate to external links: 
```

**Review by delbaoliveira — CHANGES_REQUESTED** [review]

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx (hunk 7 above)
Comment (manovotny) [reply docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:255]: Is this necessary? I don't think this page isn't surfaced in Pages router.

CleanShot 2024-01-24 at 08 52 35@2x

CleanShot 2024-01-24 at 08 52 44@2x

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx (hunk 10 above)
Comment (manovotny) [reply docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:283]:
```suggestion
In event handlers, `router.push` and `router.replace` methods form [`useRouter`](#userouter-hook) can be used to navigate to external links:
```

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 11]
```diff
@@ -252,6 +252,78 @@ export const config = {
 
 See the [Middleware](/docs/app/building-your-application/routing/middleware) documentation for more information.
 
+## External redirects
+
+The App Router fully supports programmatic redirects to external links using the same methods mentioned above.
+
+[`redirect`](#redirect-function) and [`permanentRedirect`](#permanentredirect-function) work server-side and in [Server Actions](/docs/app/building-your-application/data-fetching/server-actions-and-mutations).
```
Comment (manovotny) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:259]:
```suggestion
On the server-side and in [Server Actions](/docs/app/building-your-application/data-fetching/server-actions-and-mutations), [`redirect`](#redirect-function) and [`permanentRedirect`](#permanentredirect-function) methods can be used to navigate to external links:
```

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx (hunk 7 above)
Comment (manovotny) [reply docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:255]: Disregard my comment. I see that #61083 starts sharing the page between routers.

File: docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx [hunk 12]
```diff
@@ -252,6 +252,80 @@ export const config = {
...
+      Checkout
+    </button>
+  )
+}
+```
+
+> **Good to know**:
+>
+> - If you don't need to programmatically navigate a user, you should use a [`<Link>`](/docs/app/api-reference/components/link) component
+</AppOnly>
```
Comment (manovotny) [inline docs/02-app/01-building-your-application/01-routing/06-redirecting.mdx:328]:
```suggestion
</AppOnly>

```

**Comment by manovotny** [comment]
I have put this into draft for now.

After experimenting with both Pages and App Router, I think we can dramatically simplify this entire section.

TL;DR, you can do external redirects using any way you’d do internal redirects. As such, I don’t know that there’s much to say in a separate section aside from stating just that; *There’s nothing special about externals.*

* https://github.com/manovotny/next-app-router-redirects
* https://github.com/manovotny/next-pages-router-redirects
